### ISSUE TITLE
Many repeated and incorrect prediction lines during inference

### ISSUE THREAD

**Issue by Plaitkol** [SebastianJanampa/LINEA#5]
Your work is good. I have the problem as the title suggests when training and inference on my own dataset. 

I visualized the labels of both my own dataset and your wireframe_processed dataset. The display was correct, indicating that my dataset was formatted correctly. 

Moreover, I tested the my own test set using the following commands, and it achieved a good sAP5:
CUDA_VISIBLE_DEVICES=1 torchrun --master_port=7777 --nproc_per_node=1 main.py -c configs/linea/linea_hgnetv2_l.py --eval --resume output/checkpoint.pth --options eval_spatial_size=512 batch_size_val=8

But the mentioned problem occurs when I used the following command for inference:
python tools/inference/inference.py -c configs/linea/linea_hgnetv2_l.py -r output/checkpoint.pth -i Dataset/val -o output/inference
My inference.py is modified from your torch_inf.py. It can be used to visualize your wireframe_processed  dataset normally.

These abnormal visualizations seem to be inferred from the same image. I'm very confused about this, do you have any suggestions?

The following is part of the inference results.

<img width="2310" height="1119" alt="Image" src="https://github.com/user-attachments/assets/52d7b631-eb0f-4895-85b1-14eceaf58b5f" />

**Comment by SebastianJanampa** [comment]
Hi @Plaitkol ,
Thank you for your interest in our work. 
Could you share the inference.py file you used, please? From a quick inspection, it looks like the model is predicting the same lines across all images, with only the classification score for each line differing. Why do I say this? It's pretty interesting to see that the white part of some images has some detected lines. I would also suggest trying [HuggingFace Space](https://huggingface.co/spaces/SebasJanampa/LINEA) with the released weights first to see if the error still occurs.

**Comment by Plaitkol** [comment]
**The inference.py is:**
----------------------------------------------------------------------------------------------
"""
Copyright (c) 2024 The D-FINE Authors. All Rights Reserved.
"""
import os
import sys
import glob

import cv2  # Added for video processing
import numpy as np
import torch
import torch.nn as nn
import torchvision.transforms as T
from PIL import Image, ImageDraw

sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../../")))
from util.slconfig import SLConfig


def draw(images, lines, scores):
    for i, im in enumerate(images):
        draw = ImageDraw.Draw(im)

        scr = scores[i]
        line = lines[i][scr > thrh]
        scrs = scr[scr > thrh]

        for j, l in enumerate(line):
            draw.line(list(l), fill="orange", width=3)

            x1, y1, x2, y2 = l
            point_size = 4
            draw.ellipse([x1-point_size, y1-point_size, x1+point_size, y1+point_size], fill="#00FFFF", outline="#00FFFF")
            draw.ellipse([x2-point_size, y2-point_size, x2+point_size, y2+point_size], fill="#00FFFF", outline="#00FFFF")

            # draw.text(
            #     (l[0], l[1]),
            #     text=f"{round(scrs[j].item(), 2)}",
            #     fill="blue",
            # )

    return images 


def process_image(model, device, file_path, output_path):
    im_pil = Image.open(file_path).convert("RGB")
    w, h = im_pil.size
    orig_size = torch.tensor([[w, h]]).to(device)

    transforms = T.Compose(
        [
            T.Resize((640, 640)),
            T.ToTensor(),
            T.Normalize(mean=[0.538, 0.494, 0.453], std=[0.257, 0.263, 0.273]),
        ]
    )
    im_data = transforms(im_pil).unsqueeze(0).to(device)

    output = model(im_data, orig_size)
    lines, scores = output

    result_images = draw([im_pil], lines, scores)
    result_images[0].save(output_path)
    print(f"Image processing complete. Result saved as '{output_path}'.")


def process_video(model, device, file_path, output_path):
    cap = cv2.VideoCapture(file_path)

    # Get video properties
    fps = cap.get(cv2.CAP_PROP_FPS)
    orig_w = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
    orig_h = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))

    # Define the codec and create VideoWriter object
    fourcc = cv2.VideoWriter_fourcc(*"mp4v")
    out = cv2.VideoWriter(output_path, fourcc, fps, (orig_w, orig_h))

    transforms = T.Compose(
        [
            T.Resize((640, 640)),
            T.ToTensor(),
            T.Normalize(mean=[0.538, 0.494, 0.453], std=[0.257, 0.263, 0.273]),
        ]
    )

    frame_count = 0
    print("Processing video frames...")
    while cap.isOpened():
        ret, frame = cap.read()
        if not ret:
            break

        # Convert frame to PIL image
        frame_pil = Image.fromarray(cv2.cvtColor(frame, cv2.COLOR_BGR2RGB))

        w, h = frame_pil.size
        orig_size = torch.tensor([[w, h]]).to(device)

        im_data = transforms(frame_pil).unsqueeze(0).to(device)

        output = model(im_data, orig_size)
        lines, scores = output

        # Draw detections on the frame
        result_images = draw([frame_pil], lines, scores)

        # Convert back to OpenCV image
        frame = cv2.cvtColor(np.array(result_images[0]), cv2.COLOR_RGB2BGR)

        # Write the frame
        out.write(frame)
        frame_count += 1

        if frame_count % 10 == 0:
            print(f"Processed {frame_count} frames...")

    cap.release()
    out.release()
    print(f"Video processing complete. Result saved as '{output_path}'.")

def process_file(model, device, file_path, output_path):
    # Check if the input file is an image or a video
    if os.path.splitext(file_path)[-1].lower() in [".jpg", ".jpeg", ".png", ".bmp"]:
        # Process as image
        process_image(model, device, file_path, output_path)
    else:
        # Process as video
        process_video(model, device, file_path, output_path)

def create(args, classname):
    # we use register to maintain models from catdet6 on.
    from models.registry import MODULE_BUILD_FUNCS
    class_module = getattr(args, classname)
    assert class_module in MODULE_BUILD_FUNCS._module_dict
    build_func = MODULE_BUILD_FUNCS.get(class_module)
    return build_func(args)

def main(args):
    # Global variable
    global thrh 

    """Main function"""
    cfg = SLConfig.fromfile(args.config)

    if 'HGNetv2' in cfg.backbone:
        cfg.pretrained = False

    cfg.multiscale = None

    # build model
    model, postprocessor = create(cfg, 'modelname')

    if args.resume:
        checkpoint = torch.load(args.resume, map_location="cpu", weights_only=False)
        if "ema" in checkpoint:
            state = checkpoint["ema"]["module"]
        else:
            state = checkpoint["model"]
        model.load_state_dict(state)
    else:
        raise AttributeError("Only support resume to load model.state_dict by now.")

    class Model(nn.Module):
        def __init__(self):
            super().__init__()
            self.model = model.deploy()
            self.postprocessor = postprocessor.deploy()

        def forward(self, images, orig_target_sizes):
            outputs = self.model(images)
            outputs = self.postprocessor(outputs, orig_target_sizes)
            return outputs

    device = args.device
    model = Model().to(device)
    thrh = 0.4 if args.thrh is None else args.thrh

    # Check if the input argument is a file or a folder
    file_path = args.input
    output_path = args.output
    
    if os.path.isdir(file_path):
        # Process a folder
        folder_dir = args.input
        output_dir = args.output
        os.makedirs(output_dir, exist_ok=True)
        paths = list(glob.iglob(f"{folder_dir}/*.*"))
        for file_path in paths:
            
            filename = os.path.basename(file_path)
            name_without_ext = os.path.splitext(filename)[0]
            ext = os.path.splitext(file_path)[-1].lower()
            
            if ext in [".jpg", ".jpeg", ".png", ".bmp"]:
                output_file_path = os.path.join(output_dir, f"{name_without_ext}.jpg")
            else:
                output_file_path = os.path.join(output_dir, f"{name_without_ext}.mp4")
                
            process_file(model, device, file_path, output_file_path)
    else:
        # Process a single file
        process_file(model, device, file_path, output_path)


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser()
    parser.add_argument("-c", "--config", type=str, required=False, default='configs/linea/linea_hgnetv2_l.py')
    parser.add_argument("-r", "--resume", type=str, required=False, default='output/linea_hgnetv2_l_include/checkpoint.pth')
    parser.add_argument("-d", "--device", type=str, default="cuda:1")
    parser.add_argument("-i", "--input", type=str, required=True, help="Input file or folder path")
    parser.add_argument("-o", "--output", type=str, required=True, help="Output file or folder path")
    parser.add_argument("-t", "--thrh", type=float, required=False, default=None)
    args = parser.parse_args()
    main(args)
----------------------------------------------------------------------------------------------

**I tested the image on HuggingFace, and the result was normal:**

<img width="2214" height="1152" alt="Image" src="https://github.com/user-attachments/assets/157d2471-c2e8-4992-9d7f-dc5c696fde20" />

**Moreover, I retrained LINEA on your wireframe_processed dataset and tested it on my own dataset, and the results were still normal:**

<img width="945" height="942" alt="Image" src="https://github.com/user-attachments/assets/8d0f029f-a6bd-4c43-8e3f-937c81248f53" />

**Interestingly, I fine-tuned LINEA using my own dataset and weights that had been pre-trained on the wireframe_processed dataset, and I still encountered the same inference problem.** 

**Does this mean there's a problem with my data preparation?  However, I used the following code to visualize the labels of the wireframe_processed dataset and my own dataset：**
----------------------------------------------------------------------------------------------
import os
import json
from PIL import Image, ImageDraw
import argparse
import math
import matplotlib.pyplot as plt
import numpy as np

def convert_centered_to_pixel(coords, img_width, img_height):
    x1, y1, dx, dy = coords
    return [x1, y1, x1+dx, y1+dy]

def visualize_annotations(image_folder, json_path, output_folder, 
                                 line_color="red", line_width=2,
                                 point_color="blue", point_radius=5,
                                 show_ids=False, id_color="white"):

    os.makedirs(output_folder, exist_ok=True)
    
    with open(json_path) as f:
        data = json.load(f)
    

    id_to_image = {img['id']: img for img in data['images']}
    

    for img_id, img_info in id_to_image.items():
        img_path = os.path.join(image_folder, img_info['file_name'])
        output_path = os.path.join(output_folder, img_info['file_name'])
        
        try:
            img = Image.open(img_path).convert("RGB")
            width, height = img.size
        except FileNotFoundError:
            print(f"no find image: {img_path}")
            continue
            
        draw = ImageDraw.Draw(img)
        

        annotations = [ann for ann in data['annotations'] 
                      if ann['image_id'] == img_id and 'line' in ann]
        
        for ann in annotations:
            line = ann['line']
            ann_id = ann['id']
            
            x1, y1, x2, y2 = convert_centered_to_pixel(line, width, height)

            draw.line([(x1, y1), (x2, y2)], fill=line_color, width=line_width)

            
            circle_radius = 3
            draw.ellipse([(x1-circle_radius, y1-circle_radius), 
                          (x1+circle_radius, y1+circle_radius)], 
                         fill="aqua")
            draw.ellipse([(x2-circle_radius, y2-circle_radius), 
                          (x2+circle_radius, y2+circle_radius)], 
                         fill="aqua")
        
        
        img.save(output_path)
        print(f"Save result path: {output_path}")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="vis line label")
    parser.add_argument("--image_folder", required=True, help="input image file")
    parser.add_argument("--json_path", required=True, help="input json file")
    parser.add_argument("--output_folder", required=True, help="output path")
    parser.add_argument("--line_color", default="orange", help="line color")
    parser.add_argument("--line_width", type=int, default=3, help="line width")
    
    args = parser.parse_args()
    
    visualize_annotations(
        image_folder=args.image_folder,
        json_path=args.json_path,
        output_folder=args.output_folder,
        line_color=args.line_color,
        line_width=args.line_width
    )
----------------------------------------------------------------------------------------------

**The results were normal:**

<img width="708" height="1065" alt="Image" src="https://github.com/user-attachments/assets/b591fa4f-f7d3-457c-a0e0-a1af2a71303d" />

<img width="972" height="963" alt="Image" src="https://github.com/user-attachments/assets/9c91864a-471f-40db-9f0d-420edb2fcff0" />

This is so weird. Please help me check where the problem is. If you need any other information, let me know.

**Comment by SebastianJanampa** [comment]
Hi @Plaitkol ,
Sorry for the late response. After checking your inference.py file, i think that the only change you did was in the draw_image function. Am i right?
```python
def draw(images, lines, scores):
    for i, im in enumerate(images):
    draw = ImageDraw.Draw(im)

        scr = scores[i]
        line = lines[i][scr > thrh]
        scrs = scr[scr > thrh]

        for j, l in enumerate(line):
            draw.line(list(l), fill="orange", width=3)

            x1, y1, x2, y2 = l
            point_size = 4
            draw.ellipse([x1-point_size, y1-point_size, x1+point_size, y1+point_size], fill="#00FFFF", outline="#00FFFF")
            draw.ellipse([x2-point_size, y2-point_size, x2+point_size, y2+point_size], fill="#00FFFF", outline="#00FFFF")

            # draw.text(
            #     (l[0], l[1]),
            #     text=f"{round(scrs[j].item(), 2)}",
            #     fill="blue",
            # )

    return images 
```

If that is the case, I do not think the problem is the model. It may be your dataset. Some ideas of what's going on are that the training JSON file is wrong. Why do I think that? It is quite unusual that for all the images, you have received the same lines. So i would visualize my training dataset one by one (with augmentation techniques applied).  i have a small to visualize the augmented training set here:
https://github.com/SebastianJanampa/LINEA/blob/475c5ceea64114a48495c15888094e12f1a2d267/datasets/coco.py#L189-L233

**Comment by SebastianJanampa** [comment]
Also, did you modify the main.py file?
you said you used this code for training
```
CUDA_VISIBLE_DEVICES=1 torchrun --master_port=7777 --nproc_per_node=1 main.py -c configs/linea/linea_hgnetv2_l.py --eval --resume output/checkpoint.pth --options eval_spatial_size=512 batch_size_val=8
```
but you do not have the `coco_path` argument

**Comment by Plaitkol** [comment]
I've found the cause of the problem! I repeatedly merged the line coordinates multiple times during the JSON data preparation. I've now achieved the correct inference results. Thank you for your time and effort in helping me.

**Comment by SebastianJanampa** [comment]
I am glad you have been able to solve the problem
